### DISCUSSION TITLE
Looking for Insights on how to debug this ExtendedUpdate Error...

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#131] [Q&A]
I'm working on duplicating the example character controller into my own library.

For the most part everything looks correct and inline so I started trying to tick it. 

I duplicated the examples `prePhysicsUpdate()` call including this call:
```js
character.ExtendedUpdate(deltaTime,
	character.GetUp(),
	updateSettings,
	movingBPFilter,
	movingLayerFilter,
	bodyFilter,
	shapeFilter,
	jolt.GetTempAllocator());
```
I've checked (as far as I can in JS anyway) and all of these objects existsand the deltaTime is correctly ticking.

However, I'm getting an error:
```js
0072f336:0x3768f Uncaught RuntimeError: null function or function signature mismatch
    at 0072f336:0x3768f
    at 0072f336:0x36bed
    at 0072f336:0x198035
    at b._emscripten_bind_CharacterVirtual_ExtendedUpdate_8 (chunk-6KYJ3DMK.js?v=852e1d9b:884:22460)
    at o$.ExtendedUpdate (chunk-6KYJ3DMK.js?v=852e1d9b:16110:7)
    at characterControllerSystem.prePhysicsUpdate (character-controller.ts:418:24)
```
With the final line pointing to this: `    call_indirect (param i32 i32)`

I can't determine what's wrong from this error code and am unsure how I can get more details. I'm going to try the emscripten debugging mentioned in the readme.

**Comment by LeXXik** [comment]
Hmm. It works for me. Some notes:

- I call it the same amount of times I step the world.
- I use the same dt that I use to step the physics.
- I call it just before I step the world.
- I allocate `updateSettings` only once, and share among all chars.
- I update char ground velocity after `ExtendedUpdate`, but before I step the physics.

Try to use the same filters for a raycast and see if you get any errors.

**Reply by DennisSmolek** [reply]
I do almost everything you mentioned. 
I created an array of pre-step listeners, then fire those before the world step. They are all getting the same DeltaTime.
`ExtendedUpdateSettings` is created in the class constructor and also only once.

Right now, I'm not updating the character velocity AT ALL and getting the error. (I update the velocity in response to an input)

However, for testing this, I manually set the `linearVelocity` to `(0,0,0)` however I think this was BEFORE the update call. I'll try a version with it after.

**Reply by LeXXik** [reply]
How do you create `movingBPFilter` and `movingLayerFilter`?

**Reply by DennisSmolek** [reply]
I do it once in the class constructor: 
```js
this.objectVsBroadPhaseLayerFilter = this.joltInterface.GetObjectVsBroadPhaseLayerFilter();
this.objectLayerPairFilter = this.joltInterface.GetObjectLayerPairFilter();
this.movingBPFilter = new Jolt.DefaultBroadPhaseLayerFilter(this.objectVsBroadPhaseLayerFilter, Layer.MOVING );
this.movingLayerFilter = new Jolt.DefaultObjectLayerFilter( this.objectLayerPairFilter, Layer.MOVING );
```

**Reply by LeXXik** [reply]
Hmm, maybe you could temporarily create new ones and destroy right after, e.g.:

```js
const objectVsBroadPhaseLayerFilter = joltInterface.GetObjectVsBroadPhaseLayerFilter();
const objectLayerPairFilter = joltInterface.GetObjectLayerPairFilter();
const movingBPFilter = new Jolt.DefaultBroadPhaseLayerFilter(objectVsBroadPhaseLayerFilter, 1);
const movingLayerFilter = new Jolt.DefaultObjectLayerFilter(objectLayerPairFilter, 1);

// do extended update

Jolt.destroy(movingBPFilter);
Jolt.destroy(movingLayerFilter);
```

Also make sure `Layer.MOVING` is not `undefined` or some number that has no related filter layer.

**Reply by DennisSmolek** [reply]
Yeah I think I'll disable the loop and try just triggering the update manually with fresh ones.

**Comment by DennisSmolek** [comment]
I tried building the debug version of the library but I'm getting a Cmake/Emscripten error:
```
CMake Error at /usr/share/cmake-3.22/Modules/CMakeDetermineSystem.cmake:130 (message):
  Could not find toolchain file:
  /upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake
Call Stack (most recent call first):
  Build/Debug/_deps/joltphysics-src/Build/CMakeLists.txt:3 (project)
  ```

Which is weird because I dont get the error when building with `./build.sh Distribution`
I've also confirmed the toolchain file is in that path in the emssdk..

**Reply by DennisSmolek** [reply]
Ok I got this worked out. I had to modify and add the path in `CMakeCache.txt` in the `Build/Debug` folder

**Comment by DennisSmolek** [comment]
Got the emscripten debugger working (mostly)
It was showing me globals/variables at breakpoints but it's finicky. 
I'm getting a `memory out of bounds` error at:
Line 453 of CharacterVirtual.h
`return RMat44::sRotationTranslation(inRotation, inPosition).PreTranslated(mShapeOffset + inShape->GetCenterOfMass()).PostTranslated(mCharacterPadding * mUp);`

I was seeing that `inShape` was null. The original error message was complaining about a null function or something, although now its a memory bounds error.. 

I can't copy out the stacktrace (yet) but here's a picture:
![image](https://github.com/jrouwe/JoltPhysics.js/assets/1397052/60cf0173-db3e-49b6-928a-c4dc7420af99)

**Reply by LeXXik** [reply]
Something with the character creation settings, perhaps? You can add a guard for the shape creation, something like:
```js
const shapeResult = shapeSettings.Create();
if (shapeResult.HasError()) {
    console.error(`Failed to create shape: ${ shapeResult.GetError().c_str() }`);
    return;
}
const shape = shapeResult.Get();
characterSettings.mShape = shape;
```

**Reply by DennisSmolek** [reply]
Oh wow I didn't even know you could do a guard like that. 
I had seen `character.SetShape()` returns a boolean you could test against but I haven't seen that shape creating could fail.

**Reply by jrouwe** [reply]
If you provide sane data, shape creation can never fail. It can fail e.g. when you try to create a mesh with degenerate triangles or a box with a convex radius that's bigger than the edge length / 2.

**Comment by DennisSmolek** [comment]
Sigh....
The `inShape` error lead me on the correct cause of the problem.

I was correctly calling the `initShape()` which has that switch for Capsule, Cylinder, Box. 
However, my standing/crouching shapes were always `undefined`
turns out I accidently set the `shapeType` to `capsule` instead of uppercase `Capsule` so the switch never initialized the shapes.


This error is because there is no `Shape` on the `Character` and when it tries do do a collision detection it tries to run `inShape->GetCenterOfMass()` on a Shape that doesn't exist.


Took a few hours to figure out, but the plus side is I fixed it, got the build pipeline fixed for debug builds, and got my environment setup so I can debug both the JS and CPP parts of Jolt..
